### PR TITLE
Enable keeping standby agents

### PR DESCRIPTION
**Pull request by francoiscampbell**
Fixes https://github.com/jenkinsci/amazon-ecs-plugin/issues/93.

This PR adds a new setting, "Retain agents" to ECS Clouds. This setting (false by default), disables agent termination and keeps them around until "Container Cleanup Timeout" minutes have elapsed.

This setting is useful when running quick jobs, where the agent provisioning time can take longer than the job's own runtime.

One caution when using "Retain agents" is to ensure that the workspace is properly cleaned up in between builds. This is noted in the help for the option.

Testing this:
- Have a cloud with the agent retention off
- Run a job that uses a label from that cloud
- Wait for an agent to be started and accept the job
- Abort the job (or wait for it to complete)
- The agent is terminated
- Enable agent retention in the cloud
- Run a job that uses a label from that cloud
- Wait for an agent to be started and accept the job
- Abort the job (or wait for it to complete)
- The agent is not terminated
- Run a job that uses a label from that cloud
- The job is immediately assigned to the idle agent without provisioning a new one
- Abort the job (or wait for it to complete)
- Wait "Container Cleanup Timeout" minutes
- The agent is terminated

### CODE REVIEW

File: src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSTaskTemplate.java [hunk 1]
```diff
@@ -70,6 +70,11 @@
     @CheckForNull
     private final String label;
 
+    /**
+     * Number of nodes to retain on standby
+     */
+    private final int minRetainedNodes;
```
Comment (francoiscampbell) [inline src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSTaskTemplate.java]: this setting gets persisted per-agent, so it inherits the number at the time the agent was launched. this causes different agents to honour/not honour the actual threshold. must look for a better way to configure this, maybe a https://javadoc.jenkins.io/hudson/model/labels/LabelAtomProperty.html?

File: src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSCloud.java [hunk 2]
```diff
@@ -31,7 +31,6 @@
 import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
-import java.util.Set;
```
Comment (francoiscampbell) [inline src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSCloud.java]: this import was unused

**Comment by wuillaum** [comment]
Could the retention setting be set per template instead of a cloud default setting? It seems like locking the ability to do slave retention per cloud is a bit restrictive.

**Comment by francoiscampbell** [comment]
It could be, with the caveat that agents that are already launched when the setting is changed won't be updated, since the `RetentionStrategy` is set when the agent is launched.

My view was that if a user wants to segregate their agents by retained vs. non-retained, they could put them in different Clouds. Also, the agent cleanup timeout is on the Cloud, so it seemed logical to put agent retention on there too.

**Comment by wuillaum** [comment]
After playing around with configuration I realized there's no issue with having multiple clouds configured in Jenkins that reference the same ECS cluster. Your approach works well with this in mind.

**Comment by owengo** [comment]
Thank you for your patch! I am using it on our instance and it makes a real difference.

**Comment by webratz** [comment]
hey,
i somehow missed this. could you please rebase so we can get it merged in soon :)

**Comment by francoiscampbell** [comment]
Hey @webratz, the rebase is done.

**Pull request merged**